### PR TITLE
keep forbidden pickup/dropoff on stay in

### PR DESCRIPTION
**Pull request by antoine-de**
To keep to data as clean as possible, the model was sanitized to add restrictions on drop off at first stop time / pick up at last stop time.
Since for stay in this could remove valid journeys, the pick up at last stop time / drop off at first stop time was possible for stay in.

The previous implementation was removing explicitly set constraints on those stay in stop times.
As a side effect we now also correctly handle restrictions on stay in of different calendars (see [example 4](https://github.com/antoine-de/navitia_model/blob/fae185abe7d36cee8648ee45799853220466abd1/src/model.rs#L787-L796))

Note 1: the new implementation is quadratic in term of trip on the same `block_id`. However the number of vj on the same block_id seems quite limited thus the runtime penalty is quite small.
I tested it on [the bretagne dataset](https://transport.data.gouv.fr/datasets/base-de-donnees-multimodale-transports-publics-en-bretagne-mobibreizh/), so biggest dataset I know with block_id (106k trips, 2M+ stoptimes, max number of vehicle journeys with same block_id: 76) there is a small impact on the run time import (not on the memory footprint).
On my computer the best of 3 runs (there are ~1s of difference on the same runs) before: 17s to read the GTFS, after: 18s to read the GTFS.

Note 2:
this reactivate the use of the model_builder and add some utilities to it

internally linked to ND-1088

### CODE REVIEW

**Comment by antoine-de** [comment]
I should definitly have split this PR in 2, one for the builder and one for the bug.

I can still do it if you prefer.

I'd be happy to have a review on the builder's API

File: src/model.rs
Comment (pbougue) [inline src/model.rs:818]: If one is past-midnight and the next is at the beginning of the day after, I guess it won't work but it "only" matters for night-bus lines
:see_no_evil: becomes a bit too complex for the gain I think.

File: src/model.rs
Comment (pbougue) [inline src/model.rs:814]: Shouldn't you test it after to spare half of those overlaps checks? (not expecting any runtime optimization to compensate that)

File: src/model.rs
Comment (pbougue) [inline src/model.rs:845]: Not sure how twisted this example is (all vjs are on the same block_id, same calendar):
vj1 ends at spA (10:00-11:00)
vj2 starts at spA (10:00-11:00) ends at spB (12:00-13:00)
vj3 starts at spC (14:00-15:00)

Because of vj1-vj3 test, the outcome from your code is that all vj1's last dropoff is allowed at spA which is not what we want, no?
(we would have the same for vj2's first pickup if we had a vj0).

This probably implies building a chain first (tree if you consider calendars), then check what's possible... then decide what matters more in case of conflict (probably being safe and forbid).
And it's actually what's done in Navitia as the next vj is unique and check is done once it's affected.

File: src/model.rs
Comment (pbougue) [inline src/model.rs:853]: Can't you iterate on the &idx right away?

File: src/model.rs
Comment (pbougue) [inline src/model.rs:856]: It's uglier but for perf I think you should get that only if you really mutate it.

**Review by pbougue — COMMENTED** [review]
Did not review the model-builder part (vote for a separate PR, as we are not in a hurry to merge any of this work).

:+1: Despite remarks on twisted cases, it's better than what we had on staying as it respects data more.

File: src/model.rs
Comment (antoine-de) [reply src/model.rs:845]: hum interesting edge case, I don't know if it's a valid GTFS 🤔
I'm not sure it's worth implementing something this tricky as this is meant only for strengthening the data if we do in `transit_model` I don't think we should be "safe and forbid" (like we should in navitia) since we are only adding additional constraints on the data. So we should just forbid impossible cases, and if a particular case can happen sometimes but not always, we should let it be possible and be handled by the journey planner

File: src/model.rs
Comment (antoine-de) [reply src/model.rs:856]: it's just an access to en element by its position in a vector no?

File: src/model.rs
Comment (antoine-de) [reply src/model.rs:853]: nop since I need to mutate them, I don't see a better way to do this since there is no ìter_mut` for the moment 😞

File: src/model.rs
Comment (woshilapin) [inline src/model.rs:828]: I don't understand why you're using a `.fold()` here since you don't gather the result back in a variable. I had to pull the branch to check more, and it was really misleading. I believe a `for-loop` or at least a `.for_each()` would be more idiomatic (with a `HashMap` declared above to maintain the temporary state). And if you're worried that declaring this temporary state that won't be dropped once the processing is done, then you can extract all of that into a sub-function (so the temporary state will be dropped at the end of this sub-function).

File: src/model.rs
Comment (woshilapin) [inline src/model.rs:856]: You can consume it, it's not going to be used anymore.

```rust
       for vj_idx in vj_idxs {
            let mut vj = self.vehicle_journeys.index_mut(vj_idx);
```

**Review by woshilapin — CHANGES_REQUESTED** [review]
Overall, it's an improvement since we keep existing values and there is more test. But I feel like there is still a bunch of work to do to make it maintainable.

File: src/model.rs
Comment (antoine-de) [reply src/model.rs:1857]: I did not want to do it too early in case the api of the builder did not suit you. I think it can be done in another PR

File: src/model.rs
Comment (antoine-de) [reply src/model.rs:1867]: It's a matter of preference. I think the canonical api modeling should be as simple as possible.
And I don't really see a case where you do not want to set the stop point and (departure or arrival) so this could lead to not correctly defined stoptimes.
I think we could even make the canonical path easier if we had an easy way to set the departure value to the arrival, but I don't see a nice way to do this :/

File: src/model.rs
Comment (antoine-de) [reply src/model.rs:828]: yes I do agree that the fold does not particularly help, I can replace this by a for loop 👍

File: model-builder/src/builder.rs
Comment (antoine-de) [reply model-builder/src/builder.rs:474]: since it's in a test, I felt I should use the real name to show that it was a calendar like the others. But I'm not really think it helps, so if you prefer I can put the static variable instead

File: src/model.rs
Comment (antoine-de) [reply src/model.rs:799]: I think it was because I added some types while coding to help. you're right I can delete them 👍

File: src/model.rs
Comment (pbougue) [reply src/model.rs:829]: `_` or `String` is not helping much indeed.
Adding a block-id type may be nice, but I don't know how much it would add to this PR (and if we should explore crates to help typing more).
A comment (or a good name if you refactor to a for-loop) may be a good compromise?

File: src/model.rs
Comment (antoine-de) [reply src/model.rs:805]: hum the problem is that we cannot put a comment on the specific ìf` to explain that it handles the example 1 :/
I'm not sure it will help the understanding (I did not understand this ìf` guard immediatly in the previous implementation)

File: src/model.rs
Comment (antoine-de) [reply src/model.rs:1938]: Like I said to @pbougue I think that here we are modifying the data, so if something can happen, we have to let it be possible.
The user of the data (the journey planner for example) should be the one responsible to use a stay-in if the 2 bus are running the same days.

File: src/model.rs
Comment (pbougue) [reply src/model.rs:845]: I think this should be split into 2 parts :
1. Is it a realistic case to have chained blocks (so every extremity would endup being qualified as a valid stay-in in this algo) ? I think we get block_id corresponding to vehicles, so yes (so an algo without determining "real-life"chains would not change data much from original).
2. Is it realistic to have calendars that would make "real-life" chaining of trips being stay-in some days and not stay-in some other days? I think we should not care about that until we have the problem as it's too edgy :see_no_evil: . And OK to stay closer to data in that conflicting case, you make a good point contradicting my "safe" argument.

File: src/model.rs
Comment (woshilapin) [reply src/model.rs:1867]: I'm just worried that in some cases, canonical means something and some other cases, we discover that canonical could be something different (no departure and arrival for example, if you only care about the path and therefore the stop points)... Let's keep it like this for now and will improve over time.

File: model-builder/src/builder.rs
Comment (woshilapin) [reply model-builder/src/builder.rs:474]: Never sure what position to adopt for this kind of case, but I do get your point. I'll let you or others decide. Please resolve when you're satisfied.

File: src/model.rs
Comment (woshilapin) [reply src/model.rs:829]: This could just be creating a few aliases. Here are some ideas (not necessarily all of them has to be used, and also, naming might just be wrong!).

```rust
type VehicleJourneyIndex = (Idx<VehicleJourney>, &VehicleJourney);
type BlockVehicleJourneys = Vec<VehicleJourneyIndex>;
type BlockId = String;
type Blocks = std::collections::HashMap<BlockId, BlockVehicleJourneys>;
```

I'm mostly looking for ease of maintenance and readability here, not necessarily for strong typing.

File: src/model.rs
Comment (pbougue) [reply src/model.rs:856]: `RefMut` does a copy of `Idx` and a ref to collection (cheap) and of the `old_id`: https://docs.rs/typed_index_collection/1.2.1/src/typed_index_collection/collection.rs.html#1002-1006
And the `drop()` does some check: https://docs.rs/typed_index_collection/1.2.1/src/typed_index_collection/collection.rs.html#1018-1031

Looks like the interface on mutating CollectionWithId is not super straightforward (especially knowing that it's not zero-cost or minimal-cost).

File: src/model.rs
Comment (woshilapin) [reply src/model.rs:805]: To me, this is the level of nesting in this whole function that is making it hard to read and understand. As for the comment, it might be ok and become something like this.

> We can discard when the stop points are identicals (see Example 1 above) or when there is no stop point

Because basically, if there is only 1 stop point, then they're first and last are going to be equal.

File: src/model.rs
Comment (antoine-de) [reply src/model.rs:856]: the only 'expensive' stuff here is the string and the cost should be quite limited, I think modifying code to bypass this is really premature optimisation

File: src/model.rs
Comment (antoine-de) [reply src/model.rs]: I’m not sure its clearer but I can change it if you insist

Comment (woshilapin) [reply src/model.rs]: It would avoid having twice the default value `false` with this chaining of `.map_or()`.

Also, there is 4 different combinations (2 options) and only one is interesting and the `match` shows this better to me :smile:

Comment (antoine-de) [reply src/model.rs]: related to https://github.com/CanalTP/transit_model/pull/730#discussion_r556360803 I think one date is enough

File: src/model.rs
Comment (woshilapin) [reply src/model.rs:1938]: I had to dig a bit more to understand your point but ok, now I get it, that is the best sensible option. So I'm resolving this comment.

File: src/model.rs
Comment (pbougue) [reply src/model.rs:845]: As discussed:
1. Not addressing that for now, so authorize a bit too much pickup/dropoff which is OK (still better to respect data). To be documented but that's all.
2. If calendars introduce a contradictory case (stay-in should be authorized on sunday, not on saturday) : authorise to respect data. To be tested.

**Comment by ArnaudOggy** [comment]
To bump in `0.34` and its sub-crates :wink:

File: src/model.rs
Comment (pbougue) [inline src/model.rs:1792]: Don't know if it's better, but for oneliner const:
```suggestion
            let dates: std::collections::BTreeSet::new() = [Date::from_ymd(2020, 1, 1)].iter().cloned().collect();
```

File: src/model.rs
Comment (pbougue) [inline src/model.rs:1972]: Just an idea (resolve comment if you're not :star_struck:): it could be nice to handle iterable to do
```suggestion
                .calendar("c1", ["2020-01-01", "2020-01-02", "2020-01-03", "2020-01-04"])
```

**Review by pbougue — COMMENTED** [review]
ModelBuilder interface with default attributes + muter looks fine, and to me we settled on something good for functionnal stay-in part :+1: 
Not sure about the fallible IntoDate, and didn't look at all @woshilapin's unresolved comments, but that's the only review left before approval.

File: model-builder/src/builder.rs
Comment (antoine-de) [reply model-builder/src/builder.rs:40]: I feel this needs to be in the global comments since it's quite important, but i can also add it in the VJBuilder

File: model-builder/src/builder.rs
Comment (antoine-de) [reply model-builder/src/builder.rs:261]: I don't really know, it's a try only in the builder, and it's handy not to care about the fail.
The alternative is a `TryInto` and `d.try_into().unwrap()` in the `calendar()` method no?

File: src/model.rs
Comment (woshilapin) [reply src/model.rs:1792]: Nice one, but I'm not sure how to avoid additional allocation. Because with your solution, you have `.cloned()` which means the `Date` will exist twice (once on the stack, and once on the heap). And an alternative solution I had will allocate a `Vec`.

```rust
let dates: std::collections::BTreeSet<_> = vec![Date::from_ymd(2020, 1, 1)].into_iter().collect();
```
In the end, the solution from @antoine-de is probably the most efficient one. In one of my side project, I also did the following (but that need an additional macro).

```rust
macro_rules! btree_set {
    ($($e:expr),*) => {{
        let mut btree_set = std::collections::BTreeSet::new();
        $(btree_set.insert($e);)*
        btree_set
    }}
}
// ...
let dates = btree_set![Date::from_ymd(2020, 1, 1)];
```

But adding a new macro if it's used only once...

File: model-builder/src/builder.rs
Comment (woshilapin) [reply model-builder/src/builder.rs:261]: That wouldn't solve the problem @pbougue is trying to solve, the hidden trap if we were to use this code anywhere else than for tests. I guess the _panic_ in tests is fine, but we need to take care to these kinds of things if we were to start using it in production code... (note of the insisting guy, it would be obvious if it was named `TestModelBuilder` :wink: ).

Could we add this `IntoDate` trait with a `#[cfg(test)]` clause before (with comments about why), so it doesn't exist unless it's tests? I guess that would protect us? I would make `.calendar()` unusable in production code for now but that the goal, right?

Comment (antoine-de) [reply model-builder/src/builder.rs:261]: how about only protecting the `impl IntoDate for &str`? I'm not sure about hiding the whole `ModelBuilder` behind a `cfg(test)` (and I don't know how it will affect the use of the model builder outside of transit-model :thinking: 

Just for you information, there is the same `panic` for `IntoTime` used by `.st("stop1", "10:10:10", "10:10:11")`

Comment (woshilapin) [reply model-builder/src/builder.rs:261]: I was talking of hiding only the trait `IntoDate` but hiding only the implementation for `&str` should be enough indeed! :+1:

File: src/objects.rs
Comment (antoine-de) [reply src/objects.rs:1291]: I think it brings semantics to the operation and it will helps understand the code.
I'm not sure it really help to inline this (I don't really know the compiler policy on inlining), but if you think it's worth it, we can force the inlining here

Comment (woshilapin) [reply src/objects.rs:1291]: You're right for the 'semantic' part and I do agree with it. I'm resolving.

File: src/model.rs
Comment (antoine-de) [reply src/model.rs:2023]: oups you're right , I removed the explicit restrictions

File: src/model.rs
Comment (antoine-de) [reply src/model.rs]: after giving it some thoughts, past midnight are usually implemented as time > 24h (like `"25:00:00"` for 1am) no? and this case is handled no :thinking: ?

File: src/model.rs
Comment (antoine-de) [reply src/model.rs]: test added `block_id_on_non_overlaping_calendar_with_overlaping_stops` (and I think the current implementation handles it quite nicely after testing it :stuck_out_tongue_winking_eye: )

File: src/model.rs
Comment (antoine-de) [reply src/model.rs]: I only added the blockId type, tell me if that's ok for you

File: src/model.rs
Comment (antoine-de) [reply src/model.rs:1792]: a date is basicaly an `i32`, copying is very cheap (and we don't care much about perf here I think)
I think we should be driven by expressiveness here, I'd vote for the moment the one liner as think it's quite readable, and if we have to do it too many time, we add the macro, is that ok for you?

Comment (woshilapin) [reply src/model.rs:1792]: It's ok for me, you can replace the `.cloned()` with a `.copied()` to better express it :wink:

File: model-builder/src/builder.rs
Comment (antoine-de) [reply model-builder/src/builder.rs:257]: grumpf, does not work out of the box, are you sure it's important :roll_eyes: :smile: ?

Comment (antoine-de) [reply model-builder/src/builder.rs:257]: the other way would be to try using `const` expr on this to fail at build time, but it does not seems trivial :/

**Comment by antoine-de** [comment]
I think I resolved all issues, feel free to poke me again if I missed something

File: src/model.rs
Comment (pbougue) [reply src/model.rs]: It would not be handled I think, as the first VJ would end at `25:00` and the second would start at `02:00`.
They would be possible stay-in, but we wouldn't "see" that.
But that's OK for now (and possibly for ever :stuck_out_tongue_closed_eyes: )
I do not resolve the comment just to have it quick if we ever come back later to that work to improve those tricky cases.

**Review by pbougue — DISMISSED** [review]
Looks like all is addressed indeed (let @woshilapin check his last opened comments, though)

File: src/model.rs
Comment (pbougue) [reply src/model.rs]: Indeed, it's nice : what's consistent is handle, and what's not stays close to initial data :+1: 
Letting the discussion unresolved for the "chaining" part if we later come back to that work, but you did all that could reasonably be done :heavy_check_mark:

**Review by woshilapin — CHANGES_REQUESTED** [review]
Still need to bump `Cargo.toml` to `0.33.1` at least (or `0.34.0` as @ArnaudOggy suggested, but I would say this is bug fix so `0.33.1` is probably enough). The rest of the PR is fine for me.

**Comment by ArnaudOggy** [comment]
> Still need to bump `Cargo.toml` to `0.33.1` at least (or `0.34.0` as @ArnaudOggy suggested, but I would say this is bug fix so `0.33.1` is probably enough). The rest of the PR is good-to-go for me.

No need `0.34.0` with the re-entry of X? I let you decide what is best :wink:

**Review by pbougue — APPROVED** [review]

**Review by woshilapin — APPROVED** [review]

**Pull request merged**
